metals/src/main/scala/scala/meta/internal/metals/DocumentSymbolProvider.scala

```diff
@@ -0,0 +1,151 @@
+package scala.meta.internal.metals
+
+import scala.meta._
+import org.eclipse.lsp4j.SymbolKind
+import org.eclipse.lsp4j.DocumentSymbol
+import MetalsEnrichments._
+import scala.collection.concurrent.TrieMap
+
+/**
+ *  Retrieves all the symbols defined in a document
+ *
+ *  If the document doesn't parse, we fall back to the latest
+ *  known snapshot of the document, if present
+ *
+ *  To avoid the snapshots to grow undefinitely, we only store
+ *  the ones for the most recently requested documents
+ */
+class DocumentSymbolProvider(buffers: Buffers) {
+
+  private val snapshots: TrieMap[AbsolutePath, List[DocumentSymbol]] =
+    TrieMap.empty
+
+  def empty: List[DocumentSymbol] = Nil
+
+  def documentSymbols(
+      path: AbsolutePath
+  ): List[DocumentSymbol] = {
+    path
+      .toInputFromBuffers(buffers)
+      .parse[Source]
+      .toOption
+      .map { source =>
+        val result = new SymbolTraverser().apply(source)
+        snapshots.put(path, result)
+        result
+      }
+      .orElse(snapshots.get(path))
+      .getOrElse(empty)
+  }
+
+  def discardSnapshot(path: AbsolutePath): Unit =
+    snapshots.remove(path)
+
+  private class SymbolTraverser() {
+
+    def apply(tree: Tree): List[DocumentSymbol] = {
+      traverser.apply(tree)
+      builder.result()
+    }
+
+    private val builder = List.newBuilder[DocumentSymbol]
+
+    val traverser = new Traverser {
+      var currentRoot: Option[Tree] = None
+      override def apply(currentNode: Tree): Unit = {
+        def continue(withNewRoot: Boolean = false): Unit = {
+          val oldRoot = currentRoot
+          if (withNewRoot) currentRoot = Some(currentNode)
+          super.apply(currentNode)
+          currentRoot = oldRoot
+        }
+
+        def addName(name: String): Unit = {
+          builder += new DocumentSymbol(
+            name,
+            symbolKind(currentNode),
+            currentNode.pos.toLSP,
+            currentNode.pos.toLSP
+          )
```

Comment:
Is `currentNode.pos` the range of the whole definition (with the RHS)? I guess the 4th parameter is the `selectionRange`, which should correspond to the (more narrow) range of the identifier. This is not important, just saying that this can be improved later.

Also notice that there are more optional fields in this new `DocumentSymbol` type: `detail: String`, `deprecated: Boolean` and (the most interesting) `children: DocumentSymbol[]`. The old LSP's `SymbolInformation` declared that the range nesting and the `containerName` shouldn't be relied on for re-inferring the hierarchy, but editors used ranges anyway. Now there is this `children` field which makes this hierarchy explicit.

Reply:
Yes, I've taken some shortcuts in order to get this merged, but these are the things I would like to address next (in this order of priority):

- fallback to `SymbolInformation` for editors which don't support `DocumentSymbol` yet (e.g. vim-lsc)
- `children`
- `selectionRange`

I've agreed with @olafurpg to address those in future PR(s)

Reply:
I updated the PR so that

- `range` encloses the full range of the definition
- `selectionRange` encloses the definition name only.
- `detail` now includes the type ascription of the method (syntax written in source). It doesn't seem like vscode uses `detail` on the other hand.


We can add `deprecated` in a followup PR.

Reply:
Cool! 
I think that nested ranges would still work for `SymbolInformation` in those editors that don't support `DocumentSymbol` yet.

Reply:
Yeah, that's very much up to the editors.

For instance, the Code Outline extension for VSCode (which predates the native Outline View) used to use ranges for this. (VSCode's native view doesn't)